identity/gcp-gce-auth/README.md

```diff
@@ -0,0 +1,187 @@
+> 1. Vault authorizes the confirmed instance against the given role, ensuring the instance matches the bound zones, regions, or instance groups. If that is successful, a Vault token with the proper policies is returned.
+> https://www.vaultproject.io/docs/auth/gcp.html#gce-login
+
+<img src="https://raw.githubusercontent.com/hashicorp/vault-guides/master/assets/vault_gcp_gce_arch.png" alt="GCP gce authentication" width="400">
+
+## Background
+
+The project will create the following resources in GCP:
+
+* vault-gcp-demo-[abc] - A project with a random generated name to assosiate all the resources to
+* Vault-server - A GCE instance running Vault
+* Vault-happy - A GCE instance in the bound region - This should be able to get a token from Vault
+* Vault-sad - A GCE instance not in the bound region - This should not be able to get a token from Vault
+* vault-auth-checker - A Service Account with the Compute Viewer and Security Viewer permissions - Vault will use these credentials for the GCE backend
+
+Note: These resources will be access bound to the IP of the machine running Terraform.
+
+We will then create the following Vault install and configuration using the vault-server:
+
+* An initalized vault instance
+* One secret in the K/V store under `secret/demo`
+* A policy for access to that one secret called `reader` with permissions to read that secret
+* An auth policy tied to that reader policy that uses GCP IAM tied to a `bound_region`
+
+The initial token on the instance is created with a script on the machine stored under `/root/.vault_credentials`:
+
+```shell
+function set_vault_credentials {
+  VAULT_ADDR=${vault_addr}
+
+  JWT=\$(curl -H "Metadata-Flavor: Google"\
+  -G \
+  --data-urlencode "audience=$VAULT_ADDR/vault/web"\
+  --data-urlencode "format=full" \
+  "http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/identity")
+
+  check_errors=\$(curl \
+    --request POST \
+    --data "{\"role\": \"web\", \"jwt\": \"\$JWT\"}" \
+    "${vault_addr}/v1/auth/gcp/login" | jq -r ".errors")
+
+  if [ "\$check_errors" == "null" ]
+  then
+    VAULT_TOKEN=\$(curl \
+    --request POST \
+    --data "{\"role\": \"web\", \"jwt\": \"\$JWT\"}" \
+    "${vault_addr}/v1/auth/gcp/login" | jq -r ".auth.client_token")
+  else
+    echo "Error from vault: \$check_errors"
+    exit 1
+  fi
+
+  export VAULT_ADDR
+  export VAULT_TOKEN
+}
+```
+
+## Reference Material
+https://www.vaultproject.io/docs/auth/gcp.html
+https://www.vaultproject.io/api/auth/gcp/index.html
+
+## Note
+The code in this repository is for reference only. It is meant to illustrate a few of the requirements for using the GCP gce authentication method.
+
+## Instructions
+
+Install gcloud
+
+With Brew:
+
+```
+brew install gcloud
+```
+
+Or with the installer:
+
+```
+curl https://sdk.cloud.google.com |
+exec -l $SHELL
+gcloud init
+```
+
+Configure authentication:
+
+```
+gcloud auth login
+gcloud auth application-default login
+```
```

Comment:
I still get an authorization error during `terraform apply`:
```
Error: Error applying plan:

1 error(s) occurred:

* google_project.vault_gcp_demo: 1 error(s) occurred:

* google_project.vault_gcp_demo: Error creating project vault-gcp-demo-68528d20 (vault-gcp-demo): googleapi: Error 403: User is not authorized., forbidden.
```
Are there additional steps that're missing in the instruction?  I found [this](https://github.com/hashicorp/terraform/issues/13109).

Reply:
@yhyakuna You free to do a screenshare to go through this?

Reply:
The error may be due to my account setting.